### PR TITLE
Issue #95 - Added way for staff to uniquely identify vehicles

### PR DESCRIPTION
**Pull request by Cobchise**
#### Features: ####
1. **Scaffold** for items i.e. carts
2. Item model **delegates** to item_type model for some properties
2. Actions and views for **creating new carts** independent of model in API which are then pulled to local db
2. Ability to **edit** existing carts
3. **Implemented cart names throughout probable-engine** including scheduling rental etc
3. **Migrated Items into database** with unique name, item_type_id, and uuid
4. Implemented **'permanent-records'** for **'soft-deletion'** of items

### CODE REVIEW

File: app/controllers/items_controller.rb [hunk 1]
```diff
@@ -0,0 +1,79 @@
+class ItemsController < ApplicationController
+  before_action :set_item, only: [:show, :edit, :update]
+
+  def index
+    @item_types = ItemType.all
+    @q = Item.all.where(deleted_at: nil).search(params[:q])
```
Comment (phantompatrol) [inline app/controllers/items_controller.rb]: The `where` will likely have to become a [default scope](http://api.rubyonrails.org/classes/ActiveRecord/Scoping/Default/ClassMethods.html#method-i-default_scope)

Comment (phantompatrol) [inline app/controllers/items_controller.rb]: Also, no need for the `all` 😄

File: app/controllers/items_controller.rb [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+        flash[:success] = 'Your cart has been successfully created. '
+      end
+    else
+      flash[:danger] = 'Enter a name for the cart'
+    end
+    refresh_items
+  end
+
+  def refresh_items
+    type_id = 1
```
Comment (phantompatrol) [inline app/controllers/items_controller.rb]: ?

File: app/controllers/rentals_controller.rb [hunk 3]
```diff
@@ -109,6 +110,10 @@ def set_item_types
     @item_types = @item_types.where(name: params['item_type']).order(name: :asc) if params['item_type']
   end
 
+  def set_items
+    @items = Item.all.where(deleted_at: nil)
```
Comment (phantompatrol) [inline app/controllers/rentals_controller.rb]: No need for the `all` here

File: app/models/item.rb [hunk 4]
```diff
@@ -0,0 +1,7 @@
+class Item < ActiveRecord::Base
+  belongs_to :item_type
+  validates :name, :item_type_id, presence: true
+  validates :item_type_id, numericality: { greater_than_or_equal_to: 0 }
```
Comment (phantompatrol) [inline app/models/item.rb]: Are you able to `validate :item_type` as opposed to having to double check it, presence and numericality?

File: coverage/.last_run.json [hunk 5]
```diff
@@ -1,5 +1,5 @@
 {
   "result": {
-    "covered_percent": 95.55
+    "covered_percent": 94.89
```
Comment (phantompatrol) [inline coverage/.last_run.json]: Ehh, let's check your `coverage/index.html` to find some additional testing targets

File: db/migrate/20160715184902_create_items.rb [hunk 6]
```diff
@@ -0,0 +1,15 @@
+class CreateItems < ActiveRecord::Migration
+  def change
+    create_table :items do |t|
+    	t.string :name, unique:true, null: false
+    	t.string :item_type_id, null: false
+    	t.string :uuid, null: false
+      t.datetime :deleted_at
+      t.timestamps null: false
```
Comment (phantompatrol) [inline db/migrate/20160715184902_create_items.rb]: Spacing nit

File: db/seeds.rb [hunk 7]
```diff
@@ -49,14 +49,24 @@
 
   # Could have done in the above loop, but that was getting a bit messy
   puts 'Creating Items in the API, by Item Type'
+  items = YAML::load_file(File.join(Rails.root, 'db/db_yml', 'items.yml'))
+  itemsint = 0
```
Comment (phantompatrol) [inline db/seeds.rb]: 1. Please switch to snake_case in the future
2. Do you need this? The `i` should work, no?

File: db/seeds.rb [hunk 8]
```diff
@@ -49,14 +49,24 @@
...
+        Inventory.create_item(item_type.uuid, items[itemsint]["name"], true, {})
+        itemsint += 1
       end
     end
   end
 
+  puts 'Pulling Items from Api and Storing Locally'
+  ItemType.all.each do |item_type|
+    items = Inventory.items_by_type(item_type.uuid)
+    items.each do |item|
```
Comment (phantompatrol) [inline db/seeds.rb]: Do have enough char space (style wise) to put this on one line `Inventory.items_by_type(item_type.uuid).each do |item|`?

**Comment by phantompatrol** [comment]
This is off to a great start!

File: app/controllers/items_controller.rb [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+    @item_types = ItemType.all
+  end
+
+  def create_item
+    name = params[:name]
+    type = params[:type]
+    itemtype = ItemType.where(name: type)
+    if name != ''
+      itemtype.each do |itype|
+        Inventory.create_item(itype.uuid, name, true, {})
```
Comment (phantompatrol) [inline app/controllers/items_controller.rb]: Devil's advocate question, what happens if this fails?

File: app/views/rentals/index.html.erb [hunk 10]
```diff
@@ -16,9 +16,12 @@
   <%= f.label :item_type_id_eq, 'Item Type' %>
   <%= f.select :item_type_id_eq, options_from_collection_for_select(@item_types.order(name: :asc), "id", "name", selected: (params[:q].try(:[], :item_type_id_eq) || 0)), {include_blank: true}, class: "selectpicker form-control col-xs-2" %>
 
+  <%= f.label :item_id_eq, 'Item Name' %>
+  <%= f.select :item_id_eq, options_from_collection_for_select(@items.order(name: :asc), "id", "name", selected: (params[:q].try(:[], :item_id_eq) || 0)), {include_blank: true}, class: "selectpicker form-control col-xs-2" %>
+
   <%= f.label :user_id_eq, 'Customer' %>
   <%= f.select :user_id_eq, options_from_collection_for_select(@users.order(first_name: :asc), "id", "full_name", selected: (params[:q].try(:[], :user_id_eq) || 0)), {include_blank: true}, class: "selectpicker form-control col-xs-2" %>
-
+<br><br>
```
Comment (phantompatrol) [inline app/views/rentals/index.html.erb]: Great opportunity to play around with CSS, as opposed to line breaks

File: db/migrate/20160217192112_create_rentals.rb [hunk 11]
```diff
@@ -6,6 +6,7 @@ def change
       t.integer :department_id, null: false
       t.integer :reservation_id, null: false
       t.references :item_type, null: false
+      t.references :item, null: false
```
Comment (phantompatrol) [inline db/migrate/20160217192112_create_rentals.rb]: ⚠️  This is worth discussing! Changing old migrations can have some pretty serious ramifications! Luckily, we are not in production, but 'future you' will high five your 'past self' for knowing it.

File: spec/controllers/items_controller_spec.rb [hunk 12]
```diff
@@ -0,0 +1,95 @@
+require 'rails_helper'
+
+RSpec.describe ItemsController, type: :controller do
+
+	let!(:item) { create(:item) }
```
Comment (phantompatrol) [inline spec/controllers/items_controller_spec.rb]: Spacing nit, tabs V spaces

**Comment by phantompatrol** [comment]
Looking good!

File: app/views/rentals/index.html.erb (hunk 10 above)
Comment (phantompatrol) [inline app/views/rentals/index.html.erb]: Would you like to take a stab at this CSS before moving on (hint hint: say yes!)

Comment (Cobchise) [inline app/views/rentals/index.html.erb]: Will do! @phantompatrol

**Pull request merged**